File: android/common/src/main/java/cz/arnal/bleedge/core/Trace.kt
package cz.arnal.bleedge.core

import com.upokecenter.cbor.CBORObject
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.SecureRandom

const val TRACE_HASH_WIDTH_1 = 1
const val TRACE_HASH_WIDTH_2 = 2
const val TRACE_HASH_WIDTH_4 = 4
const val TRACE_HASH_WIDTH_8 = 8

const val TRACE_METRIC_UNKNOWN = "unknown"
const val TRACE_METRIC_RSSI = "rssi"
const val TRACE_METRIC_SNR = "snr"

data class TracePayload(
    val tag: Int,
    val authCode: Int,
    val flags: Byte,
    val routeData: ByteArray = ByteArray(0),
) {
    fun hashWidth(): Int = 1 shl (flags.toInt() and 0x03)
}

fun randomTraceTag(): Int {
    val b = ByteArray(4)
    SecureRandom().nextBytes(b)
    return ByteBuffer.wrap(b).order(ByteOrder.LITTLE_ENDIAN).int
}

fun traceFlagsForHashWidth(width: Int): Byte = when (width) {
    TRACE_HASH_WIDTH_1 -> 0
    TRACE_HASH_WIDTH_2 -> 1
    TRACE_HASH_WIDTH_4 -> 2
    TRACE_HASH_WIDTH_8 -> 3
    else -> error("invalid trace hash width $width")
}.toByte()

fun encodeTracePayload(p: TracePayload): ByteArray {
    val out = ByteBuffer.allocate(9 + p.routeData.size).order(ByteOrder.LITTLE_ENDIAN)
    out.putInt(p.tag)
    out.putInt(p.authCode)
    out.put(p.flags)
    out.put(p.routeData)
    return out.array()
}

fun decodeTracePayload(payload: ByteArray): TracePayload {
    require(payload.size >= 9) { "trace payload too short: ${payload.size}" }
    val buf = ByteBuffer.wrap(payload).order(ByteOrder.LITTLE_ENDIAN)
    val tag = buf.int
    val auth = buf.int
    val flags = buf.get()
    val routeData = ByteArray(buf.remaining())
    buf.get(routeData)
    return TracePayload(tag, auth, flags, routeData)
}

fun traceRouteData(route: List<NodeID>, width: Int = TRACE_HASH_WIDTH_8): ByteArray {
    traceFlagsForHashWidth(width)
    val out = ByteArray(route.size * width)
    route.forEachIndexed { i, id ->
        id.bytes.copyInto(out, destinationOffset = i * width, startIndex = 0, endIndex = width)
    }
    return out
}

data class TraceResult(
    val tag: Int,
    val authCode: Int,
    val route: List<NodeID>,
    val forwardNodes: List<NodeID>,
    val forwardSamples: List<Byte>,
    val returnNodes: List<NodeID> = emptyList(),
    val returnSamples: List<Byte> = emptyList(),
    val metric: String = TRACE_METRIC_RSSI,
) {
    fun encode(): ByteArray {
        val map = CBORObject.NewOrderedMap()
        map[CBORObject.FromObject(1)] = CBORObject.FromObject(tag)
        map[CBORObject.FromObject(2)] = CBORObject.FromObject(authCode)
        val routeArr = CBORObject.NewArray()
        route.forEach { routeArr.Add(CBORObject.FromObject(it.bytes)) }
        map[CBORObject.FromObject(3)] = routeArr
        val fNodeArr = CBORObject.NewArray()
        forwardNodes.forEach { fNodeArr.Add(CBORObject.FromObject(it.bytes)) }
        map[CBORObject.FromObject(4)] = fNodeArr
        val fSampleArr = CBORObject.NewArray()
        forwardSamples.forEach { fSampleArr.Add(CBORObject.FromObject(it.toInt())) }
        map[CBORObject.FromObject(5)] = fSampleArr
        if (returnNodes.isNotEmpty()) {
            val rNodeArr = CBORObject.NewArray()
            returnNodes.forEach { rNodeArr.Add(CBORObject.FromObject(it.bytes)) }
            map[CBORObject.FromObject(6)] = rNodeArr
        }
        if (returnSamples.isNotEmpty()) {
            val rSampleArr = CBORObject.NewArray()
            returnSamples.forEach { rSampleArr.Add(CBORObject.FromObject(it.toInt())) }
            map[CBORObject.FromObject(7)] = rSampleArr
        }
        map[CBORObject.FromObject(8)] = CBORObject.FromObject(metric)
        return map.EncodeToBytes()
    }

    companion object {
        fun decode(payload: ByteArray): TraceResult {
            val map = CBORObject.DecodeFromBytes(payload)
            return TraceResult(
                tag = map[CBORObject.FromObject(1)].AsInt32(),
                authCode = map[CBORObject.FromObject(2)].AsInt32(),
                route = cborArrayToNodeIDs(map[CBORObject.FromObject(3)]),
                forwardNodes = cborArrayToNodeIDs(map[CBORObject.FromObject(4)]),
                forwardSamples = cborArrayToBytes(map[CBORObject.FromObject(5)]),
                returnNodes = cborArrayToNodeIDs(map[CBORObject.FromObject(6)]),
                returnSamples = cborArrayToBytes(map[CBORObject.FromObject(7)]),
                metric = map[CBORObject.FromObject(8)]?.takeIf { !it.isNull }?.AsString() ?: TRACE_METRIC_UNKNOWN,
            )
        }
    }
}

File: core/trace.go
package core

import (
	"encoding/binary"
	"fmt"

	"github.com/fxamacker/cbor/v2"
)

const (
	TraceHashWidth1 = 1
	TraceHashWidth2 = 2
	TraceHashWidth4 = 4
	TraceHashWidth8 = 8

	TraceMetricUnknown = "unknown"
	TraceMetricRSSI    = "rssi"
	TraceMetricSNR     = "snr"
)

// TracePayload follows MeshCore's TRACE payload prefix:
// [tag:4 LE][auth_code:4 LE][flags:1][route_hashes...].
//
// flags & 0x03 selects each route hash width: 0=1B, 1=2B, 2=4B, 3=8B.
// BLEEdge currently uses 8-byte NodeID route hashes for exact routing.
type TracePayload struct {
	Tag       uint32
	AuthCode  uint32
	Flags     byte
	RouteData []byte
}

func TraceFlagsForHashWidth(width int) (byte, error) {
	switch width {
	case TraceHashWidth1:
		return 0, nil
	case TraceHashWidth2:
		return 1, nil
	case TraceHashWidth4:
		return 2, nil
	case TraceHashWidth8:
		return 3, nil
	default:
		return 0, fmt.Errorf("invalid trace hash width %d", width)
	}
}

func (p TracePayload) HashWidth() int {
	return 1 << (p.Flags & 0x03)
}

func (p TracePayload) RouteHashes() [][]byte {
	width := p.HashWidth()
	if width <= 0 || len(p.RouteData) == 0 {
		return nil
	}
	n := len(p.RouteData) / width
	out := make([][]byte, n)
	for i := range out {
		out[i] = append([]byte(nil), p.RouteData[i*width:(i+1)*width]...)
	}
	return out
}

func EncodeTracePayload(p TracePayload) []byte {
	payload := make([]byte, 9+len(p.RouteData))
	binary.LittleEndian.PutUint32(payload[0:4], p.Tag)
	binary.LittleEndian.PutUint32(payload[4:8], p.AuthCode)
	payload[8] = p.Flags
	copy(payload[9:], p.RouteData)
	return payload
}

func DecodeTracePayload(payload []byte) (TracePayload, error) {
	if len(payload) < 9 {
		return TracePayload{}, fmt.Errorf("trace payload too short: %d", len(payload))
	}
	return TracePayload{
		Tag:       binary.LittleEndian.Uint32(payload[0:4]),
		AuthCode:  binary.LittleEndian.Uint32(payload[4:8]),
		Flags:     payload[8],
		RouteData: append([]byte(nil), payload[9:]...),
	}, nil
}

func TraceRouteData(route []NodeID, width int) ([]byte, error) {
	if _, err := TraceFlagsForHashWidth(width); err != nil {
		return nil, err
	}
	out := make([]byte, 0, len(route)*width)
	for _, id := range route {
		out = append(out, id[:width]...)
	}
	return out, nil
}

// TraceResult is returned to the originator when a trace reaches its destination.
// ForwardSamples are collected on the outbound request. ReturnSamples are collected
// on the result packet as it travels back. On macOS CoreBluetooth exposes RSSI but
// not LoRa SNR, so Metric is currently "rssi"; future radios can use "snr".
type TraceResult struct {
	Tag            uint32   `cbor:"1,keyasint"`
	AuthCode       uint32   `cbor:"2,keyasint"`
	Route          []NodeID `cbor:"3,keyasint"`
	ForwardNodes   []NodeID `cbor:"4,keyasint"`
	ForwardSamples []int8   `cbor:"5,keyasint"`
	ReturnNodes    []NodeID `cbor:"6,keyasint,omitempty"`
	ReturnSamples  []int8   `cbor:"7,keyasint,omitempty"`
	Metric         string   `cbor:"8,keyasint"`
}

type traceResultWire struct {
	Tag            int32    `cbor:"1,keyasint"`
	AuthCode       int32    `cbor:"2,keyasint"`
	Route          []NodeID `cbor:"3,keyasint"`
	ForwardNodes   []NodeID `cbor:"4,keyasint"`
	ForwardSamples []int8   `cbor:"5,keyasint"`
	ReturnNodes    []NodeID `cbor:"6,keyasint,omitempty"`
	ReturnSamples  []int8   `cbor:"7,keyasint,omitempty"`
	Metric         string   `cbor:"8,keyasint"`
}

func (r TraceResult) Encode() ([]byte, error) {
	return cbor.Marshal(traceResultWire{
		Tag:            int32(r.Tag),
		AuthCode:       int32(r.AuthCode),
		Route:          r.Route,
		ForwardNodes:   r.ForwardNodes,
		ForwardSamples: r.ForwardSamples,
		ReturnNodes:    r.ReturnNodes,
		ReturnSamples:  r.ReturnSamples,
		Metric:         r.Metric,
	})
}

func DecodeTraceResult(payload []byte) (TraceResult, error) {
	var raw map[int]cbor.RawMessage
	if err := cbor.Unmarshal(payload, &raw); err != nil {
		return TraceResult{}, err
	}
	tag, err := decodeTraceUint32(raw[1])
	if err != nil {
		return TraceResult{}, fmt.Errorf("trace tag: %w", err)
	}
	auth, err := decodeTraceUint32(raw[2])
	if err != nil {
		return TraceResult{}, fmt.Errorf("trace auth: %w", err)
	}
	var w traceResultWire
	w.Tag = int32(tag)
	w.AuthCode = int32(auth)
	if err := decodeTraceField(raw, 3, &w.Route); err != nil {
		return TraceResult{}, fmt.Errorf("trace route: %w", err)
	}
	if err := decodeTraceField(raw, 4, &w.ForwardNodes); err != nil {
		return TraceResult{}, fmt.Errorf("trace forward nodes: %w", err)
	}
	if err := decodeTraceField(raw, 5, &w.ForwardSamples); err != nil {
		return TraceResult{}, fmt.Errorf("trace forward samples: %w", err)
	}
	if err := decodeTraceField(raw, 6, &w.ReturnNodes); err != nil {
		return TraceResult{}, fmt.Errorf("trace return nodes: %w", err)
	}
	if err := decodeTraceField(raw, 7, &w.ReturnSamples); err != nil {
		return TraceResult{}, fmt.Errorf("trace return samples: %w", err)
	}
	if err := decodeTraceField(raw, 8, &w.Metric); err != nil {
		return TraceResult{}, fmt.Errorf("trace metric: %w", err)
	}
	return TraceResult{
		Tag:            uint32(w.Tag),
		AuthCode:       uint32(w.AuthCode),
		Route:          w.Route,
		ForwardNodes:   w.ForwardNodes,
		ForwardSamples: w.ForwardSamples,
		ReturnNodes:    w.ReturnNodes,
		ReturnSamples:  w.ReturnSamples,
		Metric:         w.Metric,
	}, nil
}

func decodeTraceUint32(raw cbor.RawMessage) (uint32, error) {
	if len(raw) == 0 {
		return 0, nil
	}
	var v int64
	if err := cbor.Unmarshal(raw, &v); err != nil {
		return 0, err
	}
	return uint32(v), nil
}

func decodeTraceField(raw map[int]cbor.RawMessage, key int, dst any) error {
	msg := raw[key]
	if len(msg) == 0 {
		return nil
	}
	return cbor.Unmarshal(msg, dst)
}
